src/server/bg_services/replication_log_parser.js

```diff
@@ -0,0 +1,320 @@
+/* Copyright (C) 2016 NooBaa */
+'use strict';
+
+const AWS = require('aws-sdk');
+const system_store = require('../system_services/system_store').get_instance();
+const replication_store = require('../system_services/replication_store').instance();
+const cloud_utils = require('../../util/cloud_utils');
+const http_utils = require('../../util/http_utils');
+const pool_server = require('../system_services/pool_server');
+const dbg = require('../../util/debug_module')(__filename);
+const config = require('../../../config');
+const moment = require('moment');
+
+/**
+ * get_log_candidates will return an object which contains the log candidates
+ * @param {*} source_bucket_id ID of the source bucket
+ * @param {String} rule_id ID of the replication rule
+ * @param {Record<any, any>} replicationconfig Replication configuration
+ * @param {Number} candidates_limit Candidates limit
+ * @param {(log_entry: Record<string, any>) => Boolean} ignore_fn - function that returns true if the log entry should be ignored
+ * @returns 
+ */
+async function get_log_candidates(source_bucket_id, rule_id, replicationconfig, candidates_limit, ignore_fn = () => false) {
+    if (replicationconfig && replicationconfig.log_replication_info && replicationconfig.log_replication_info.aws_log_replication_info) {
+        return get_aws_log_candidates(source_bucket_id, rule_id, replicationconfig, candidates_limit, ignore_fn);
+    }
+}
+
+async function get_aws_log_candidates(source_bucket_id, rule_id, replication_config, candidates_limit, ignore_fn) {
+    let aws_log_replication_info = replication_config.log_replication_info.aws_log_replication_info;
+    const { logs_bucket, prefix } = aws_log_replication_info.logs_location;
+    let continuation_token = get_continuation_token_for_rule(rule_id, replication_config);
+    const s3 = get_source_bucket_aws_connection(source_bucket_id, aws_log_replication_info);
+
+    let candidates_obj = {};
+    let logs_retrieved_count = config.AWS_LOG_CANDIDATES_LIMIT;
+
+    do {
+        const next_log_entry = await aws_get_next_log_entry(s3, logs_bucket, prefix, continuation_token);
+
+        // save the continuation token for the next iteration
+        continuation_token = next_log_entry.NextContinuationToken;
+
+        // Check if there is any log entry - if there are no more log entries
+        // then no need to process anything
+        if (!next_log_entry.Contents || next_log_entry.Contents.length === 0) {
+            dbg.log0('log_parser: no more log entries');
+            break;
+        }
+
+        const next_log_data = await aws_get_next_log(s3, logs_bucket, next_log_entry.Contents[0].Key);
+        aws_parse_log_object(candidates_obj, next_log_data, ignore_fn);
+
+        logs_retrieved_count -= 1;
```

Comment:
I don't understand the use of `logs_retrieved_count`, why do you change it and why is it inside the while loop?

Reply:
It is to ensure that only `AWS_LOG_CANDIDATES_LIMIT` logs are processed in a single iteration.

Reply:
Marking resolved - addressed.